feat(noRestrictedImports): add the patterns option

sakai-ast opened the pull request:
<!--
	Thanks for submitting a Pull Request! We appreciate you spending the time to work on these changes.
	Please provide enough information so that others can review your PR.
	Once created, your PR will be automatically labeled according to changed files.
	Learn more about contributing: https://github.com/biomejs/biome/blob/main/CONTRIBUTING.md
-->

## Summary
Implements: https://eslint.org/docs/latest/rules/no-restricted-imports#patterns
closes: https://github.com/biomejs/biome/issues/5289

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve?-->

<!-- Link any relevant issues if necessary or include a transcript of any Discord discussion. -->

## Test Plan
documentation and snapshot tests

<!-- What demonstrates that your implementation is correct? -->

ematipico commented on crates/biome_js_analyze/Cargo.toml:
```diff
@@ -34,7 +34,9 @@ biome_unicode_table      = { workspace = true }
 bitvec                   = "1.0.1"
 camino                   = { workspace = true }
 enumflags2               = { workspace = true }
+fancy-regex              = "0.14.0"
```
Do we really need yet another crate for regex? We already have `regex` in the workspace.

sakai-ast replied:
Yes, when I used the `regex` crate, I got the following error.
```rust
called `Result::unwrap()` on an `Err` value: Syntax(
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
regex parse error:
    import-foo/(?!bar)
               ^^^
error: look-around, including look-ahead and look-behind, is not supported
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

If we don't need the look-around feature, I use `regex` craete.

siketyan replied:
They are not implemented in the `regex` crate for performance reasons, and I don't think these fancy features are typically needed for import restrictions. Plus, we should use the `fancy-regex` crate everywhere in Biome for consistency, if it is really needed.

sakai-ast replied:
okay, I use `regex` instead.

sakai-ast replied:
https://github.com/biomejs/biome/pull/5506/commits/8f79a44e3ca51bae01fcbb7267ea6b36b6000ee1

Conaclos replied:
Actually you shroud even use our wrapper named `RestrictedRegex` that ensures use of a consistent syntax for all regexes in the configuration files. We already use this type in `useNamingConvention` and `useFilenamingConvention`. You can directly replace `Box<str> by `RestrictedRegex` where you expect a regex. This will validate and deserialize it correctly (then you will not have to compile the regex and handle errors) :)

ematipico commented on crates/biome_js_analyze/tests/specs/nursery/noRestrictedImports/invalid.js.snap:
```diff
@@ -1,7 +1,7 @@
 ---
 source: crates/biome_js_analyze/tests/spec_tests.rs
+assertion_line: 134
```
Please make sure you have the latst version of `cargo-insta`. This addition is part of an old version of the tool.

sakai-ast replied:
https://github.com/biomejs/biome/pull/5506/commits/d708d7c3e533f2b3a9f7caf1e44226a270a219cb

ematipico commented on crates/biome_js_analyze/src/lint/nursery/no_restricted_imports.rs:
```diff
@@ -490,9 +805,347 @@ impl Deserializable for CustomRestrictedImport {
+#[cfg_attr(feature = "schema", derive(schemars::JsonSchema))]
+#[serde(rename_all = "camelCase", deny_unknown_fields, default)]
+pub struct PatternOptions {
+    /// An array of gitignore-style patterns.
```
We already possess a crate called `biome_glob` for globs. We don't need to reinvent the wheel :) We don't need to use the `ignore` crate. (`biome_glob` uses `ignore` under the hood, but it imposes few restrictions).

sakai-ast replied:
I've already tried to use the `biome_glob`, but I got the following error.

```
Failed to parse group property gitignore-style pattern 'import-foo/prefix[A-Z]*': Character class `[]` are not supported. Use `\[` and `\]` to escape the characters.
```

Is it ok to use the `biome_glob` without supporting the character class `[]`?

siketyan replied:
It's okay not supporting character classes as users can use regex instead.

sakai-ast replied:
https://github.com/biomejs/biome/pull/5506/commits/1a30b72ff58fce82d239f0a31c16560af5c1fe11

ematipico commented on crates/biome_js_analyze/src/lint/nursery/no_restricted_imports.rs:
```diff
@@ -490,9 +805,348 @@ impl Deserializable for CustomRestrictedImport {
+
+impl PatternOptions {
+    // Ensure that mutually exclusive keys are not used together.
+    fn validate_combination(&self) -> Result<(), &'static str> {
```
cc @Conaclos  and @siketyan I believe you both implemented custom schema representation, maybe you can help

sakai-ast replied:
I tried implementing this here: https://github.com/biomejs/biome/pull/5506/commits/7e93bdac4d6e0a32cdaf12cc5670bef8bf457da5

ematipico commented on crates/biome_js_analyze/src/lint/nursery/no_restricted_imports.rs:
```diff
@@ -952,6 +1658,30 @@ pub struct RestrictedImportMessage {
     allowed_import_names: Box<[Box<str>]>,
 }
 
+impl RestrictedImportMessage {
```
Please move ALL the implementations after `impl Rule for ...`

We usually structure the rules like this:
- declare macro
- `impl Rule`
- the rest

sakai-ast replied:
https://github.com/biomejs/biome/pull/5506/commits/9a9ea462e31eaa0d4937927551aeb3a0134a5b1d

sakai-ast commented:
Sorry for the delay. I've completed the initial implementation and it's now ready for review.

dyc3 commented on Cargo.toml:
I'm not sure why this diff is here. Doesn't appear relevant to this PR. Can we revert the changes to this file?

ematipico replied:
Indeed. This is probably a mistake from the diff resolution

ematipico reviewed (changes requested):
Thank you @sakai-ast for the amazing work! Users will really appreciate that. I left a couple of suggestions we should address. Also, there a couple of important things we should address **first**:
- Lack of new tests: we added **many** new options, but those are hardly tested. I know we have docs, but those are mostly for showing how they work to the users. We should really have a bigger test suite for us to cover all the new options.
- We should [create a new changeset ](https://github.com/biomejs/biome/blob/main/CONTRIBUTING.md#create-a-changeset). It should be a `minor`, and it should be brief enough to at least explain a basic new option.

Conaclos commented on crates/biome_js_analyze/src/lint/style/no_restricted_imports.rs:
```diff
@@ -315,6 +341,201 @@ declare_lint_rule! {
+    /// {
+    ///     "options": {
+    ///        "patterns": [{
+    ///             "group": ["utils/*"],
```
I wonder if we should rename `group` by `globs`. This look more intuitive to me. Any opinion?

Conaclos commented on crates/biome_js_analyze/src/lint/style/no_restricted_imports.rs:
```diff
@@ -386,115 +860,440 @@ impl ImportRestrictionStatus {
-    message: Box<str>,
+pub struct PatternOptions {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    group: Option<Box<[Box<str>]>>,
```
We introduced our own glob type some months ago.
And the import organizer provides a wrapper named `ImportSourceGlob` dedicated to source glob matching.
Thus, I could write this:

```suggestion
    group: Option<Box<[ImportSourceGlob]>>,
```

Actually, I wonder if we should take one more step and use the same source matcher as our import organizer:

```suggestion
    group: Option<SourcesMatcher>,
```

This could allow users to use a glob or a list of globs.
Also, this could allow users to use predefined matchers like `:NODE:` (cf the import organizer for more details).

Going even farther, we could use `GroupMatcher` that supports matcher like `{ type: true, source: "@app/**" }`. I could extend it and add support for matching imported names (e.g. `{ type: true, names: ["A", "B"] source: "@app/**" }`). This could remove the need of options like `importNames`.
This requires some discussion with other team members.

Conaclos commented on crates/biome_js_analyze/src/lint/style/no_restricted_imports.rs:
```diff
@@ -386,115 +860,440 @@ impl ImportRestrictionStatus {
-    #[serde(skip_serializing_if = "<[_]>::is_empty")]
-    import_names: Box<[Box<str>]>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    regex: Option<Box<str>>,
```
We have our own regex type:
```suggestion
    regex: Option<RestrictedRegex>,
```

Also, I wonder if we really need this option.
We decided to go for globs when we match import sources.

Conaclos commented on crates/biome_js_analyze/src/lint/style/no_restricted_imports.rs:
```diff
@@ -386,115 +860,440 @@ impl ImportRestrictionStatus {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    allow_import_names: Option<Box<[Box<str>]>>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    import_name_pattern: Option<Box<str>>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    allow_import_name_pattern: Option<Box<str>>,
```
I wonder if we really need these two options in a first version. Do you know if it is widely sued?

Conaclos reviewed (commented):
My main remark is about using `Box<str>` instead of `Glob` and `RestrictedRegex` that could save you time and avoid custom validation and handling of errors.

I am also concerned by the number of options and the fact that some options exclude others.
I wonder if we really have to support regex matching of import paths (aka import sources). Globs could be enough.
Similarly importnames and `importnamePatterns` seems a bit of duplicates to me. I know it is for performance. However, it is perhaps early sued options? In this case we could just provide `importNamePatterns`

sakai-ast commented:
@ematipico @Conaclos @dyc3 
Thank you for your taking time to review this PR and for your valuable feedback.

Before I address each review comment, I'd like to discuss and finalize the scope of initial implementation.

My initial approach was to implement all possible ESLint patterns options. However, I now believe a more focused version based on Conaclos' suggestions would be more practical. I suggest we concentrate on the following core options:

- group (or globs, as discussed in https://github.com/biomejs/biome/pull/5506#discussion_r2101078727)
- import name patterns
  - A boolean option to invert pattern behavior(similar to the current allow options)
 
**Please note:** I think this approach would also involve removing the following patterns option.
```
"patterns": ["utils/*", "!utils/foo"]),
```
This functionality would be covered by the group option.

These changes will provide a simple set of options needed as an initial version, as well as simplify implementation.

I'm keen to hear your thoughts on this proposed direction before I proceed with the necessary code changes and address the other comments.

siketyan commented:
I agree with Conaclos' opinion at all, we can add only the most valuable option for now and leave others later for future opportunities. Plus, we don't have to copy all options from ESLint, it can be added by historical reasons which we don't need to follow.

dyc3 commented:
I think using our globs for the pattern matching is a good idea. Consistency is good. As a user, I would expect wherever I can put in a pattern, the patterns would have the same semantics.

Allowing to match on `organizeImports` groups is interesting. For example, it would allow users to ban all `:NODE:` imports, which would be useful in a vanilla js project that is meant to run in a browser.

sakai-ast commented:
Remaining Tasks
- add lack of new tests(WIP)
   - https://github.com/biomejs/biome/pull/5506#pullrequestreview-2853089856 
-  about TODO
     - https://github.com/biomejs/biome/pull/5506#discussion_r2097251791
 - create a new change set
    - https://github.com/biomejs/biome/pull/5506#pullrequestreview-2853089856

sakai-ast commented on crates/biome_js_analyze/src/lint/style/no_restricted_imports.rs:
```diff
@@ -360,19 +648,205 @@ enum ImportRestrictionCause {
+                }
+            }
+            AnyJsImportLike::JsImportCallExpression(import_call) => {
+                // TODO: We have to parse the context of the import() call to determine
```
@ematipico This TODO wasn't created by me, and I don't plan to address it in this PR (Of course I would like to resolve this in this PR, but it is taking too long, so I give up for now...). However, should I create an issue for it?
https://github.com/biomejs/biome/blob/main/crates/biome_js_analyze/src/lint/style/no_restricted_imports.rs#L1006-L1009


Another TODO also exists in the no_restricted_imports.rs. Should I also create an issue for this as well?
https://github.com/biomejs/biome/blob/main/crates/biome_js_analyze/src/lint/style/no_restricted_imports.rs#L894-L897

sakai-ast commented:
@ematipico 
Sorry for taking so long.
I have now finished the following remaining tasks.
https://github.com/biomejs/biome/pull/5506#issuecomment-3007024488

ematipico commented:
@sakai-ast we recently changed how rule options are implemented, and I moved all the new options inside the new place. Everything that you implemented should be there. Now, while I was doing that, I noticed a recursion bug that is clippy flags:

Here 

https://github.com/sakai-ast/biome/blob/e43aa5202bc6dbcc82601f2bf91b4efb068d667f/crates/biome_rule_options/src/no_restricted_imports.rs#L387-L400

And Here

https://github.com/sakai-ast/biome/blob/e43aa5202bc6dbcc82601f2bf91b4efb068d667f/crates/biome_rule_options/src/no_restricted_imports.rs#L407-L420

As you can see, the implementation of the methods calls itself, causing a recursion. I'm not sure what the correct implementation of those methods is, so I ask you to address them. As a general rule, use `cargo clippy` or `just l` to check for possible violations.

sakai-ast commented:
@ematipico 
Thank you for resolving the conflict!

I don't think this is a recursion bug - the methods eventually call the `PathOptions` and `PatternOptions` implementations, not themselves.
In my environment, the clippy flags only appear  in `impl ImportRestrictions for PatternOptions` . 
This is because the implementation is missing.

I’ve now:

- added an implementation for `PatternOptions` (This was missing during the migration)

- moved some functions into `biome_rule_options` side(for the above)

- removed some unnecessary code(This was added during the migration)

- fixed an issue where only the paths option was honored when both paths and patterns were set(I noticed this while doing this work)

With these changes, both `cargo clippy` and `just l` run without errors.

The test output has changed compared with the previous run, but before updating the snapshots I’d like to confirm that these fixes look correct to you.

ematipico commented:
Linting is passing, but it seems the tests aren't due to snapshot changes

sakai-ast commented:
@ematipico 
I had intended to wait for your confirmation for my fix before updating the snapshots, but I have now applied the snapshot fix.

dyc3 commented on crates/biome_js_analyze/src/lint/style/no_restricted_imports.rs:
````diff
@@ -315,6 +332,93 @@ declare_lint_rule! {
+    /// import { foo } from 'import-foo/foo';
+    /// ```
+    ///
+    /// ### `invertImportNamePattern`
+    ///
+    /// If true, the matched patterns in the importNamePattern will be allowed
+    ///
````
```suggestion
    /// ### `invertImportNamePattern`
    /// **Since `v2.1.0`**
    ///
    /// If true, the matched patterns in the importNamePattern will be allowed
    ///
```

dyc3 commented on crates/biome_js_analyze/src/lint/style/no_restricted_imports.rs:
````diff
@@ -315,6 +332,93 @@ declare_lint_rule! {
+    /// import bar from 'import-foo/bar';
+    /// ```
+    ///
+    /// ### `importNamePattern`
+    ///
+    /// This option allows you to use regex patterns to restrict import names.
+    ///
````
```suggestion
    /// ### `importNamePattern`
    /// **Since `v2.1.0`**
    ///
    /// This option allows you to use regex patterns to restrict import names.
    ///
```

dyc3 commented on crates/biome_js_analyze/src/lint/style/no_restricted_imports.rs:
````diff
@@ -315,6 +332,93 @@ declare_lint_rule! {
     /// import { Bar } from 'import-bar';
     /// ```
+    ///
+    /// ### `patterns`
+    ///
+    /// This option allows you to specify multiple modules to restrict using gitignore-style patterns.
````
```suggestion
    /// ### `patterns`
    /// **Since `v2.1.0`**
    ///
    /// This option allows you to specify multiple modules to restrict using gitignore-style patterns.
```

dyc3 reviewed (approved):
Looking good! Just a couple docs changes.

siketyan commented on crates/biome_js_analyze/tests/specs/style/noRestrictedImports/invalid.options.json:
```diff
@@ -66,7 +66,38 @@
 								]
 							}
-						}
+						},
```
Nit: looks tab and space indents are mixed here.

siketyan commented on crates/biome_js_analyze/tests/specs/style/noRestrictedImports/valid.options.json:
```diff
@@ -73,7 +73,38 @@
 								]
 							}
-						}
+						},
```
Ditto

siketyan commented on crates/biome_rule_options/src/no_restricted_imports.rs:
```diff
@@ -77,9 +98,201 @@ pub struct CustomRestrictedImportOptions {
     pub allow_import_names: Box<[Box<str>]>,
 }
 
+impl PathOptions {
```
I think we should avoid to put js-specific stuff in the biome_rule_options crate if possible, as it will be shared among languages.

siketyan replied:
Plus the rule implementation for a language should be colocated.

siketyan reviewed (approved):
Great work! I left some suggestions for refactoring, not a blocker at all :)

sakai-ast commented on .changeset/better-adults-roll.md:
```diff
@@ -0,0 +1,6 @@
+"@biomejs/biome": minor
+---
+
+An option called `patterns` has been added to the `noRestrictedImports` rule. This option allows you to specify multiple modules to restrict using gitignore-style patterns.
```
Thank you for your consideration.
I was able to set aside some time and I improved the changeset.
https://github.com/biomejs/biome/pull/5506/commits/c94f7a2447d88ebc5dee885630b3aca2ca9913ee

ematipico commented on .changeset/better-adults-roll.md:
````diff
@@ -0,0 +1,79 @@
+import { foo } from 'import-foo/foo';
+```
+
+### Valid examples
````
```suggestion
 **Valid examples**
```

dyc3 commented on .changeset/better-adults-roll.md:
````diff
@@ -0,0 +1,79 @@
+}
+```
+
+### Invalid examples
````
https://github.com/DavidAnson/markdownlint/blob/main/doc/md036.md

ematipico replied:
Can we turn it off?

dyc3 replied:
Yes, that would be done in this file https://github.com/biomejs/biome/blob/main/.markdownlint.json
